core/domain/user_jobs_one_off.py

```diff
@@ -150,6 +152,123 @@ def reduce(userbio_length, stringified_usernames):
+        # Mark story and topic as incomplete for incomplete explorations.
+        for story in stories_linked_to_incomplete_explorations:
+            if story:
+                learner_progress_services.mark_story_as_incomplete(
+                    user_id, story.id)
+                learner_progress_services.mark_topic_as_partially_learnt(
+                    user_id, story.corresponding_topic_id)
+
+        # Mark story and topic as incomplete for completed explorations.
+        for story in stories_linked_to_completed_explorations:
+            if story:
+                completed_nodes = (
+                    story_fetchers.get_completed_nodes_in_story(
+                        user_id, story.id))
+                all_nodes = story.story_contents.nodes
+
+                if len(completed_nodes) != len(all_nodes):
+                    learner_progress_services.mark_story_as_incomplete(
+                        user_id, story.id)
+                    learner_progress_services.mark_topic_as_partially_learnt(
+                        user_id, story.corresponding_topic_id)
+
+        yield (
+            'Successfully added story_ids and topic_ids '
+            'to the story_ids and partially_learnt_topic_ids field '
+            'of IncompleteActivitiesModel for the user: %s' % (user_id), 1)
+
+    @staticmethod
+    def reduce(key, values):
+        yield (key, len(values))
+
+
+class PopulateStoriesAndTopicsInCompletedActivitiesOneOffJob(
+        jobs.BaseMapReduceOneOffJobManager):
+    """One-off job to populate the story_ids and learnt_topic_ids
+    in Completed Activities Model."""
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        """Return a list of datastore class references to map over."""
+        return [user_models.UserSettingsModel]
+
+    @staticmethod
+    def map(item):
+        """Implements the map function for this job."""
+        user_id = item.id
+        completed_exploration_ids = (
+            learner_progress_services.get_all_completed_exp_ids(user_id))
+
+        story_ids_linked_to_completed_explorations = (
+            list(set(exp_services.get_story_ids_linked_to_explorations(
+                completed_exploration_ids))))
+        stories_linked_to_completed_explorations = (
+            story_fetchers.get_stories_by_ids(
+                story_ids_linked_to_completed_explorations))
+
+        # Mark story and topic as completed for completed explorations.
+        for story in stories_linked_to_completed_explorations:
+            if story:
+                completed_nodes = (
+                    story_fetchers.get_completed_nodes_in_story(
+                        user_id, story.id))
+                all_nodes = story.story_contents.nodes
+
+                if len(completed_nodes) == len(all_nodes):
+                    learner_progress_services.mark_story_as_completed(
+                        user_id, story.id)
+                    learner_progress_services.mark_topic_as_learnt(
```

Comment:
Same thing here (a topic is learnt if at least one story is fully learnt).

Reply:
Done. PTAL!